### PR TITLE
Use embedfs & goose for migrations

### PR DESCRIPTION
**Pull request by archseer**
[ch6124]

This drops the gorm dependency on migrations.

Migrations are now located in `core/store/migrate/migrations/`.  By default, migrations are simple `.sql` files with some comment annotations:

```sql
-- +goose Up
CREATE TABLE post (
    id int NOT NULL,
    title text,
    body text,
    PRIMARY KEY(id)
);

-- +goose Down
DROP TABLE post;
```

(More complex SQL statements also need to be surrounded `-- +goose StatementBegin` and `-- +goose StatementEnd`, mostly function calls)

This makes it easier to fix migration conflicts because only the file needs renaming.

If necessary (for instance to manually migrate some data using ORM calls), a`.go` migration can still be created. See `0036_external_job_id.go` for an example.

- `chainlink node db status` now displays a table of all migrations along with whether they're already applied (and when).
- `chainlink node db create-migration [-type go] my_migration_name` will generate a new migration file.
- `chainlink node db rollback` will roll back a single migration, `chainlink node db rollback <version>` will roll back to a specific version. 

See the [goose README](https://github.com/pressly/goose#migrations) for more details.

### CODE REVIEW

File: core/store/migrate/migrate.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
+package migrate
```
Comment (archseer) [inline core/store/migrate/migrate.go:1]: The bulk of changes are in this file, migrations just got adapted to fit the new format without any changes.

File: core/cmd/app.go [hunk 2]
```diff
@@ -620,12 +620,30 @@ func NewApp(client *Client) *cli.App {
...
+							Action: client.StatusDatabase,
+							Flags:  []cli.Flag{},
+						},
 						{
 							Name:   "migrate",
 							Usage:  "Migrate the database to the latest version.",
 							Action: client.MigrateDatabase,
 							Flags:  []cli.Flag{},
 						},
+						{
```
Comment (samsondav) [inline core/cmd/app.go:635]: Amazing, any chance of a rollback command?

File: core/cmd/local_client.go [hunk 3]
```diff
@@ -436,12 +437,67 @@ func (cli *Client) VersionDatabase(c *clipkg.Context) error {
 		return fmt.Errorf("failed to initialize orm: %v", err)
 	}
 
-	version, err := migrations.Current(orm.DB)
+	db := postgres.UnwrapGormDB(orm.DB).DB
```
Comment (samsondav) [inline core/cmd/local_client.go:440]: Eventually I think we might want to instantiate it separately but this is fine for now

File: core/cmd/local_client.go [hunk 4]
```diff
@@ -436,12 +437,67 @@ func (cli *Client) VersionDatabase(c *clipkg.Context) error {
...
+// StatusDatabase displays the database migration status
+func (cli *Client) StatusDatabase(c *clipkg.Context) error {
+	logger.SetLogger(cli.Config.CreateProductionLogger())
+	cfg := cli.Config
+	parsed := cfg.DatabaseURL()
+	if parsed.String() == "" {
+		return cli.errorOut(errors.New("You must set DATABASE_URL env variable. HINT: If you are running this to set up your local test database, try DATABASE_URL=postgresql://postgres@localhost:5432/chainlink_test?sslmode=disable"))
+	}
+
+	orm, err := orm.NewORM(parsed.String(), cfg.DatabaseTimeout(), gracefulpanic.NewSignal(), cfg.GetDatabaseDialectConfiguredOrDefault(), cfg.GetAdvisoryLockIDConfiguredOrDefault(), cfg.GlobalLockRetryInterval().Duration(), cfg.ORMMaxOpenConns(), cfg.ORMMaxIdleConns())
```
Comment (samsondav) [inline core/cmd/local_client.go:460]: This `NewORM` thing is nasty, I have this card for fixing: https://app.clubhouse.io/chainlinklabs/story/14027/pull-initialization-of-db-out-of-orm-and-inject-into-newapplication

File: core/store/migrate/migrate.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+//go:embed migrations/*.sql
+var embedMigrations embed.FS
+
+const MIGRATIONS_DIR string = "migrations"
+
+func init() {
+	goose.SetBaseFS(embedMigrations)
+	goose.SetSequential(true)
+	goose.SetTableName("goose_migrations")
+	// goose.SetVerbose(true) // can be set to debug migrations
```
Comment (samsondav) [inline core/store/migrate/migrate.go:27]: Maybe link it up to the LOG_SQL_MIGRATIONS env var?

File: core/store/migrate/migrate.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+		_, err = db.Exec("DROP TABLE migrations;")
+		return err
+	})
+	if err != nil {
+		panic(err)
+	}
+}
+
+func Migrate(db *sql.DB) error {
```
Comment (samsondav) [inline core/store/migrate/migrate.go:65]: I assume you have tested on at least one aged production database?

File: core/store/migrate/migrations/0001_initial.sql [hunk 7]
```diff
@@ -3001,8 +2995,10 @@ ALTER TABLE ONLY public.task_specs
 --
 -- PostgreSQL database dump complete
 --
-`
-	down1 = `
+
+-- +goose StatementEnd
+
+-- +goose Down
```
Comment (samsondav) [inline core/store/migrate/migrations/0001_initial.sql:3001]: Oh man, this is great

**Review by samsondav — APPROVED** [review]

File: core/cmd/local_client.go (hunk 4 above)
Comment (archseer) [reply core/cmd/local_client.go:460]: Yeah it's a whole lot of copypasta..

File: core/store/migrate/migrate.go (hunk 6 above)
Comment (archseer) [reply core/store/migrate/migrate.go:65]: I tested it on my local db with:

- An empty db (`reset`)
- A fully migrated db that used the old migrations (I ran reset on the develop branch, switched over then ran migrate)
- A partially migrated db that used the old migrations (I ran reset after deleting a couple migrations on develop, then switched over)

File: core/cmd/app.go (hunk 2 above)
Comment (archseer) [reply core/cmd/app.go:635]: Done! `chainlink node db rollback` will roll back a single migration, `chainlink node db rollback <version>` will roll back to a specific version.

**Review by samsondav — APPROVED** [review]

**Pull request merged**